### ISSUE TITLE
negative cumulative time in STDIO module

### ISSUE THREAD

**Issue by shanedsnyder** [darshan-hpc/darshan#229]
In GitLab by @carns on May 1, 2017, 11:46

Reported by Sebastian Oeste.  IOR reports negative time values for STDIO but not POSIX, and the problem seems to get worse for larger/longer IOR runs.

**Comment by shanedsnyder** [comment]
In GitLab by @carns on May 2, 2017, 10:43

[issue-229-oeste-negative-timer-dbg.patch](/uploads/fe97f55d0f86dd1f6efd1aa4a15dfddd/issue-229-oeste-negative-timer-dbg.patch)

I've asked Sebastian to try the attached patch.  It includes a minor bug fix (the READ macro is using a write variable to track the last time for overlaps), but I don't think that is causing the problem.  I'm more curious about the fprintf output to confirm if either a) the timers really are negative on a particular wrapper or b) the macro has a logic bug because both _tm1 and _tm2 are before the last value.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jul 21, 2017, 13:14

I can reproduce this one. Looking into it now, will maybe start by applying this patch and seeing if something stands out.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jul 21, 2017, 13:37

Looks like this bug only happens when running with multiple processes and doesn't affect any module but STDIO. Seems like there has to be an error in the shared reduction code for STDIO.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jul 21, 2017, 15:14

Actually scratch that previous comment, it's not a problem with the shared file reductions. The problem is non-deterministic though -- the STDIO timer values are not always negative, but some of the time (irrespective of the number of processes).

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jul 21, 2017, 16:59

Looks like the failure happens in the "bad last logic" debug printf case in Phil's patch. Somehow the STDIO module is calling DARSHAN_TIMER_INC_NO_OVERLAP in some cases where the ending timestamp of the operation (__tm2) is 0.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jul 24, 2017, 13:45

Slowly learning more about this, but it looks like the calls that result in this behavior follow this code path:

fflush wrapper  
STDIO_RECORD_WRITE macro  
DARSHAN_TIMER_INC_NO_OVERLAP macro  

I don't see how tm2 timer in fflush (i.e., fflush ending timestamp) could be set to zero, unless darshan_core_wtime itself is returning 0. darshan_core_wtime will return a zero value timer if darshan core runtime is not up and running... so, I'm guessing the flushing of stdout (or whatever stream it is) is being done *after* Darshan has already shutdown, causing the error? Need to confirm and think about how to avoid that...

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jul 25, 2017, 10:57

mentioned in merge request !15

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Jul 25, 2017, 11:13

This issue was fixed in merge request !15.

The issue was that fflush is being used to flush stdout while Darshan is in the process of shutting down. Modules are supposed to stop instrumenting once Darshan core has began to shutdown, but there hasn't been a way for modules to reliably determine this -- at the very least, Darshan core will prevent new modules/records from being registered once shutdown has began, but if a module already has a reference for a specific record then it can try to update it.

In this specific case, the STDIO module already has a reference for stdout, so attempts to update it -- but, darshan_core_wtime is returning 0 since Darshan is shutting down, leading to the negative timer values ultimately. All modules are susceptible to this problem, just seems STDIO was the first to be bitten.

I added a call for darshan core for modules to check whether shutdown has started -- if so, they skip the instrumentation step and just call the underlying I/O call.
